src/providers/yamcs-web-socket.js

```diff
@@ -0,0 +1,147 @@
+        };
+
+        this._socket.onmessage = (event) => {
+            this._onmessageSubscribers.forEach(onmessageSubscriber => onmessageSubscriber.callback(event));
+        };
+
+        this._socket.onerror = (error) => {
+            console.error(error);
+            console.warn("Websocket error, attempting reconnect...");
+            this._connected = false;
+            this._reconnect();
+
+            this.onerrorSubscribers.forEach(onerrorSubscriber => onerrorSubscriber.callback(error));
+        };
+
+        this._socket.onclose = () => {
+            this._connected = false;
+            console.warn("Websocket closed. Attempting to reconnect...");
+            this._reconnect();
+
+            this._oncloseSubscribers.forEach(oncloseSubscriber => oncloseSubscriber.callback());
+        };
+    }
+
+    onclose(callback) {
+        const oncloseCallback = { callback };
+
+        this._oncloseSubscribers.add_(oncloseCallback);
+
+        return () => {
+            this._oncloseSubscribers.delete(oncloseCallback);
+        };
+    }
+
+    onerror(callback) {
+        const onerrorCallback = { callback };
+
+        this._onerrorSubscribers.add(onerrorCallback);
+
+        return () => {
+            this._onerrorSubscribers.delete(onerrorCallback);
+        };
+    }
+
+    onmessage(callback) {
+        const onmessageCallback = { callback };
+
+        this._onmessageSubscribers.add(onmessageCallback);
+
+        return () => {
+            this._onmessageSubscribers.delete(onmessageCallback);
+        };
+    }
+
+    onopen(callback) {
+        const onopenCallback = { callback };
+
+        this._onopenSubscribers.add(onopenCallback);
+
+        return () => {
+            this._onopenSubscribers.delete(onopenCallback);
+        };
+    }
+
+    sendMessage(message) {
+        if (!this._connected) {
+            this._requests.push(message);
+
+            return;
+        }
+
+        this._sendOrQueueMessage(message);
+    }
+
+    _flushQueue() {
+        this._requests = this._requests.filter(request => {
+            try {
+                this._sendMessage(request);
+            } catch (error) {
+                console.error(error);
+
+                return true;
+            }
+
+            return false;
+        });
+    }
+
+    _reconnect() {
+        if (this._reconnectTimeout) {
+            return;
+        }
+
+        this._reconnectTimeout = setTimeout(() => {
+            this.createWebsocket();
+            delete this._reconnectTimeout;
+        }, FALLBACK_AND_WAIT_MS[this._currentWaitIndex]);
+
+        if (this._currentWaitIndex < FALLBACK_AND_WAIT_MS.length - 1) {
+            this._currentWaitIndex++;
+        }
+    }
+
+    _sendMessage(message) {
+        this._socket.send(message);
+    }
+
+    _sendOrQueueMessage(message) {
+        try {
+            this._sendMessage(message);
+        } catch (error) {
+            console.error(error);
```

Comment:
On error this should reconnect, [as in the original code](https://github.com/akhenry/openmct-yamcs/blob/7916c86ea7bac5037f8c4c5253857aad58bf3739/src/providers/realtime-provider.js#L130). As I said, __we have to get this right__. The stability of our WebSocket connections has been hard won over the years, with a lot of trial and error.